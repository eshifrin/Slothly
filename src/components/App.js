import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  ImageBackground,
  Modal,
  ScrollView
} from 'react-native';
import PropTypes from 'prop-types';
import AddSessionInput from './AddSessionInput';
import db from '../db/firebase';
import SessionList from './SessionList';
import Editor from './Editor';
import Footer from './Footer';
import RenderIf from '../utils/RenderIf';
import { calcStats } from '../utils/calcStats';
import _get from 'lodash.get';

export default class App extends Component {
  static defaultProps = {
    sessions: []
  };

  state = {
    showEditorModal: false,
    sessions: this.props.sessions,
    stats: {}
  };

  componentWillMount() {
    this.getSessionData();
  }

  componentDidCatch(error, info) {
    console.log('caught error', error, info); //TODO: add logger
  }

  addSession = sessionLength => {
    return db
      .addSession({
        sessionLength: Number(sessionLength)
      })
      .then(this.getSessionData());
  };

  deleteSession = sessionId => {
    return db.deleteSession(sessionId).then(this.getSessionData());
  };

  editSession = idx => {
    this.setState({
      indexInEditing: idx,
      showEditorModal: true
    });
  };

  closeModal = () => {
    this.setState({
      showEditorModal: false,
      indexInEditing: undefined
    });
  };

  updateSession = update => {
    return db
      .updateSession(update)
      .then(this.getSessionData())
      .then(() => this.closeModal());
  };

  getSessionData = () => {
    return db.getAllSessions().then(sessions => {
      this.setState({
        sessions,
        stats: calcStats(sessions)
      });
    });
  };

  render() {
    const sessionToEdit =
      this.state.indexInEditing !== undefined
        ? this.state.sessions[this.state.indexInEditing]
        : {};

    return (
      <View style={styles.container}>
        <Modal visible={this.state.showEditorModal} animationType="slide">
          <Editor
            sessionToEdit={sessionToEdit}
            onCancel={this.closeModal}
            updateSession={this.updateSession}
          />
        </Modal>
        <AddSessionInput addSession={this.addSession} />
        <SessionList
          sessions={this.state.sessions}
          editSession={this.editSession}
          deleteSession={this.deleteSession}
        />
        <Footer stats={this.state.stats} />
      </View>
    );
  }
}

App.propTypes = {
  sessions: PropTypes.array,
  stats: PropTypes.shape({
    minutesSoFar: PropTypes.string,
    yearlyPace: PropTypes.string
  })
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: 30
  },
  content: {
    flex: 1
  }
});
