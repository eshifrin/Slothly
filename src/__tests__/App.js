import 'react-native';
import React from 'react';
import App from '../components/App';
import { shallow } from 'enzyme';

describe('App', () => {
  let wrapper;
  beforeEach(done => {
    wrapper = shallow(<App />);
    setTimeout(() => done(), 0); //ensures lifecycle hooks are run
  });

  it('renders', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('with mock data', () => {
    expect(wrapper.state()).toMatchSnapshot();
  });

  it('can add sessions', () => {
    const randomNum = Math.floor(Math.random() * 10000);

    wrapper
      .instance()
      .addSession(randomNum)
      .then(() => {
        expect(
          wrapper
            .state()
            .sessions.map(({ sessionLength }) => sessionLength)
            .includes(randomNum)
        ).toBeTruthy;
      });
  });

  it('can delete sessions', () => {
    expect(wrapper.state().sessions.length).toBeTruthy;

    wrapper
      .instance()
      .deleteSession('mockId1')
      .then(() => {
        expect(wrapper.state().sessions.length).toBeFalsey;
      });
  });

  it('can update sessions', () => {
    const randomNum = Math.floor(Math.random() * 10000);
    const id = wrapper.state().sessions[0].id;

    wrapper
      .instance()
      .updateSession({
        id,
        sessionLength: randomNum
      })
      .then(() => {
        expect(wrapper.state().sessions[0].sessionLength).toEqual(randomNum); // ?
      });
  });
});
