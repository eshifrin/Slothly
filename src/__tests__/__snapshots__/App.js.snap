// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingTop": 30,
    }
  }
>
  <Component
    animationType="slide"
    hardwareAccelerated={false}
    visible={false}
  >
    <Editor
      onCancel={[Function]}
      sessionToEdit={Object {}}
      updateSession={[Function]}
    />
  </Component>
  <AddSessionInput
    addSession={[Function]}
  />
  <SessionList
    deleteSession={[Function]}
    editSession={[Function]}
    sessions={Array []}
  />
  <Component
    stats={Object {}}
  />
</View>
`;

exports[`App with mock data 1`] = `
Object {
  "sessions": Array [
    Object {
      "id": "mockId2",
      "note": "mock note",
      "sessionLength": 5,
      "when": 2018-01-01T08:00:00.000Z,
    },
  ],
  "showEditorModal": false,
  "stats": Object {
    "minutesSoFar": "5",
    "yearlyPace": "86",
  },
}
`;
